hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DefaultSource.scala

```diff
@@ -164,6 +164,18 @@ class DefaultSource extends RelationProvider
     new HoodieEmptyRelation(sqlContext, df.schema)
   }
 
+  /**
+   * For primary key less tables, incoming df could contain meta fields and we can't remove them. for other
+   * flows we can remove the meta fields.
+   * @param optParams
+   * @return
+   */
+  private def canRemoveMetaFields(optParams: Map[String, String]) : Boolean = {
+    !(optParams.getOrDefault(SPARK_SQL_WRITES_PREPPED_KEY, "false").toBoolean
+    || optParams.getOrDefault(SPARK_SQL_MERGE_INTO_PREPPED_KEY, "false").toBoolean
+    || !optParams.containsKey(RECORDKEY_FIELD.key()))
```

Comment:
For pk-less tables, RECORDKEY_FIELD may not be present and this condition can be true. Do we want to drop the meta fields in such cases? From the comment it seems like for pk-less we don;t want to drop the meta fields.

Reply:
not sure I get it. we will drop the meta fields down the line while creating the HoodieRecordPayload. 
here we are just trying to gauge if we can go w/ prepped writes or regular non-prepped writes. if incoming df has meta fields and if its pk less table, we go with prepped flow.

Reply:
My point was for pk-less this condition would be true and `canRemoveMetaFields` can return true, isn't it? However, we need the meta fields for pk-less.
```
!optParams.containsKey(RECORDKEY_FIELD.key()
```